src/gui/dialogs/file_dialog.cpp

```diff
@@ -0,0 +1,558 @@
+	// (in save mode non-existent files are accepted as long as the parent dir
+	// exists).
+	const std::string& absolute_parent = fs::normalize_path(new_parent, true, true);
+	if(!absolute_parent.empty()) {
+		DBG_FILEDLG << "register_new_selection(): new selection '" << name << "' does not exist or is not accessible, but parent exists\n";
+		current_dir_ = absolute_parent;
+		current_entry_ = fs::base_name(name);
+		return SELECTION_NOT_FOUND;
+	}
+
+	DBG_FILEDLG << "register_new_selection(): new selection '" << name << "' does not exist or is not accessible\n";
+	return SELECTION_PARENT_NOT_FOUND;
+}
+
+void tfile_dialog::set_input_text(ttext_box& t, const std::string& value)
+{
+	if(value.empty()) {
+		clear_input_text(t);
+		return;
+	}
+
+	t.set_value(value);
+
+	const size_t vallen = t.get_length();
+	const size_t extlen = utf8::size(extension_);
+
+	if(save_mode_ && extlen && vallen > extlen) {
+		// Highlight everything but the extension if it matches
+		if(value.substr(vallen - extlen) == extension_) {
+			t.set_selection(0, vallen - extlen);
+		}
+	}
+}
+
+void tfile_dialog::clear_input_text(ttext_box& t)
+{
+	if(save_mode_ && !extension_.empty()) {
+		t.set_value(extension_);
+		t.set_selection(0, 0);
+	} else {
+		t.clear();
+	}
+}
+
+void tfile_dialog::refresh_fileview(twindow& window)
+{
+	cursor::setter cur{cursor::WAIT};
+
+	dir_files_.clear();
+	dir_subdirs_.clear();
+
+	// TODO: Need to detect and handle cases where we don't have search permission
+	//       on current_dir_, otherwise things may get weird.
+	filesystem::get_files_in_dir(current_dir_, &dir_files_, &dir_subdirs_, filesystem::FILE_NAME_ONLY);
+	isort_dir_entries(dir_files_);
+	isort_dir_entries(dir_subdirs_);
+
+	//
+	// Clear and refill the filelist box.
+	//
+
+	tlistbox& filelist = find_widget<tlistbox>(&window, "filelist", false);
+
+	filelist.clear();
+
+	// Parent entry
+	if(!fs::is_root(current_dir_)) {
+		// label_parent may not necessarily be always ".." in the future, so push
+		// with check_selection = false and check the selection ourselves here.
+		push_fileview_row(filelist, label_parent, icon_parent, false);
+		if(current_entry_ == PARENT_DIR || current_entry_.empty()) {
+			filelist.select_row(0, true);
+		}
+	}
+
+	for(const auto& dir : dir_subdirs_) {
+		push_fileview_row(filelist, dir, icon_dir);
+	}
+
+	for(const auto& file : dir_files_) {
+		push_fileview_row(filelist, file, icon_file);
+	}
+
+	find_widget<tcontrol>(&window, "current_dir", false).set_label(current_dir_);
+	set_input_text(find_widget<ttext_box>(&window, "filename", false), current_entry_);
+}
+
+void tfile_dialog::push_fileview_row(tlistbox& filelist, const std::string& name, const std::string& icon, bool check_selection)
+{
+	// TODO: Hopefully some day GUI2 will allow us to make labels be ellipsized
+	//       dynamically at layout/rendering time.
+	std::string label = name;
+	utils::ellipsis_truncate(label, FILE_DIALOG_MAX_ENTRY_LENGTH);
```

Comment:
There is not, no, only if re-setting a label whose size has already been calculated.